naptime/src/main/scala/org/coursera/naptime/Types.scala

```diff
@@ -91,34 +92,39 @@ object Types extends StrictLogging {
       case unknown: DataSchema =>
         throw new RuntimeException(s"Cannot compute asymmetric type for key type: $unknown")
     }
-    for ((name, graphQLRelation) <- fields.graphQLRelations) {
-      Option(mergedSchema.getField(name)) match {
-        case Some(field) =>
-          logger.warn(
-            s"Fields for resource $typeName tries to add graphql relation on field " +
-              s"'$name', but that field is already defined on the model")
-        case None =>
-          val errorMessageBuilder = new StringBuilder
-          val newField = graphQLRelation.toAnnotation.relationType match {
-            case FINDER | MULTI_GET =>
-              val newField = new RecordDataSchema.Field(new ArrayDataSchema(new StringDataSchema)) // TODO(bryan): fix type here
-              newField.setOptional(false)
-              newField
-            case SINGLE_ELEMENT_FINDER | GET =>
-              val newField = new RecordDataSchema.Field(new StringDataSchema) // TODO(bryan): fix type here
-              newField.setOptional(true)
-              newField
-            case _ =>
-              throw new RuntimeException(
-                "unknown relation type: " +
-                  graphQLRelation.toAnnotation.relationType.toString)
-          }
-          val relationMap = Map[String, AnyRef](
-            Relations.RELATION_PROPERTY_NAME -> graphQLRelation.toAnnotation.data())
-          newField.setProperties(relationMap.asJava)
-          newField.setDoc(graphQLRelation.description)
-          newField.setName(name.split("/").last, errorMessageBuilder)
-          insertFieldAtLocation(mergedSchema, name.split("/").dropRight(1).toList, newField)
+    val relationKeys = fields.relations.keys ++ fields.graphQLRelations.keys
+    for (name <- relationKeys) {
+      val relationOption = fields.relations.get(name)
+      val graphQLRelationOption = fields.graphQLRelations.get(name)
+      if (mergedSchema.contains(name)) {
+        logger.warn(
+          s"Fields for resource $typeName tries to add Include/GraphQL relation on field " +
+            s"'$name', but that field is already defined on the model")
+      } else {
+        val nameSegments = name.split("/")
+        val errorMessageBuilder = new StringBuilder
+        val effectiveRelationType = graphQLRelationOption.map(_.toAnnotation.relationType)
+        val newField = effectiveRelationType match {
+          case None | Some(FINDER) | Some(MULTI_GET) =>
```

Comment:
I think the `None` case should be called out explicitly, from my reading, in that case, relatedOption must not be none.

In that case, is `new RecordDataSchema.Field(new ArrayDataSchema(new StringDataSchema))` on L109 always correct for the related resource?

Reply:
Yeah, the `relatedOption != None` is implied since we iterate over the keySet. Actually, now that I think about it more, I think this executes multiple times because I didn't union the keys, I just concatenated their iterables.

I'm going to go back on my idea here and give it a specific `null` type instead. I don't plan on using it in my type generator, and it serves no value to Assembler being the same. Plus, if the relation is 1-1, then when they add a GraphQL relation, it'll change types anyway, I'm not sure why I thought I could protect against that.